File: python/main/edrm/MappingEntry.py
import hashlib
import urllib
from datetime import datetime
from typing import Any, Union
from xml.dom.minidom import Document, Element

import edrm.EDRMUtilities as eutes
import edrm
from edrm import FieldFactory
from edrm.EntryField import EntryField
from edrm.EntryInterface import EntryInterface


class MappingEntry(EntryInterface):
    """
    Generic data entry for key:value pair mappings, such as rows in a CSV file, database, or JSON file.

    This class can, and should be subclassed when the generic means of calculating names and item dates aren't
    suitable.  Additionally, all incoming dtae/times are assumed to be instances datetime.datetime.
    """

    def __init__(self, mapping: dict[str, Any], mimetype: str, parent_id: str = None):
        super().__init__()

        self.__data: dict[str, Any] = mapping
        self.__parent_id: str = parent_id

        self.__fill_initial_fields()
        self.__fill_generic_fields(mimetype)

    def __fill_initial_fields(self):
        for key, value in self.data.items():
            if isinstance(value, bool):
                data_type = EntryField.TYPE_BOOLEAN
                _value = value
            elif isinstance(value, int):
                data_type = EntryField.TYPE_INTEGER
                _value = value
            elif isinstance(value, float):
                data_type = EntryField.TYPE_DECIMAL
                _value = value
            elif isinstance(value, datetime):
                data_type = EntryField.TYPE_DATETIME
                _value = value
            else:
                data_type = EntryField.TYPE_TEXT
                _value = str(value)

            self[key] = FieldFactory.generate_field(key, data_type, _value)

    def __fill_generic_fields(self, mimetype: str):
        self['MIME Type'] = FieldFactory.generate_field('MIME Type', EntryField.TYPE_TEXT, mimetype)
        self['SHA-1'] = FieldFactory.generate_field('SHA-1',
                                                    EntryField.TYPE_TEXT,
                                                    eutes.hash_data(self.data, hashlib.sha1()))
        self['Name'] = FieldFactory.generate_field('Name', EntryField.TYPE_TEXT, self.name)
        self['Item Date'] = FieldFactory.generate_field('Item Date', EntryField.TYPE_DATETIME, self.itemdate)

    @property
    def data(self) -> dict[str, Any]:
        return self.__data

    @property
    def identifier_field(self) -> str:
        return 'SHA-1'

    @property
    def name(self) -> str:
        """
        Attempt to locate a name in the provided dictionary of data.  If no name is present, then return the first
        value present in the data.

        :return: A string to use as the data row's display name
        """
        field_names = list(self.data.keys())
        first_field = field_names[0]
        name_fields = [f for f in field_names if 'name' in f.lower()]

        if edrm.configs['default_rowname_field'] in field_names:
            name_field = edrm.configs['default_rowname_field']
        elif len(name_fields) == 0:
            name_field = first_field
        else:
            name_field = name_fields[0]

        return str(self.data[name_field])

    @property
    def time_field(self) -> Union[str, None]:
        """
        Provide a name of a field to be used as the row items <code>Item Date</code>.  The item date is a specific
        required property used to locate data and events on a timeline.

        This method will search for time / date fields in the following order:
        * If there is a field named "CreateTime" that will be used.
        * If there is a field with the word "time" (case-insensitive) in its name, it will be used.
        * If there is a field with the word "date" (case-insensitive) in its name, it will be used.

        Barring any of those options, None will be provided.

        If a specific field which would not be found in the above order then a subclass of this class should be used
        to override this property.
        :return: A field that be used as the Item Date, or None if an appropriate field can't be found.
        """
        field_names = list(self.data.keys())

        if edrm.configs['default_itemdate_field'] in field_names:
            return edrm.configs['default_itemdate_field']

        time_fields = [f for f in field_names if 'time' in f.lower()]
        if len(time_fields) > 0:
            return time_fields[0]

        date_fields = [f for f in field_names if 'date' in f.lower()]
        if len(date_fields) > 0:
            return date_fields[0]

        return None

    @property
    def itemdate(self) -> datetime:
        """
        Attempt to lookup an item date or time for this mapping.  If there isn't one present, use "now"  This
        assumes that the values store for the field are either of the type datetime.datetime, or a string in the
        Python format "%Y-%m-%d %H:%M:%S.%f" (more commonly: YYYY-mm-DD HH:MM:SS.ssssss).  If neither of these is
        true an exception will be raised.

        :return: A time to use for this entry's timeline time.
        """
        time_field = self.time_field

        if time_field is None:
            return datetime.now()
        else:
            date_time = self.data.get(time_field)

            if isinstance(date_time, datetime):
                return date_time
            elif isinstance(date_time, str):
                return datetime.strptime(date_time, edrm.configs['date_time_format'])
            else:
                raise ValueError(f'Invalid item date format: {date_time}')

    @property
    def parent(self) -> str:
        return self.__parent_id

    def add_as_parent_path(self, existing_path: str):
        return self.name + '/' + existing_path

    def calculate_md5(self) -> str:
        return eutes.hash_data(self.data, hashlib.md5())

    def add_location_uri(self,
                         document: Document,
                         container: Element,
                         entry_map: dict[str, EntryInterface],
                         for_nli: bool):
        if for_nli:
            location_uri_element = document.createElement('LocationURI')
            location_uri = eutes.generate_relative_path(self, entry_map)
            location_uri = urllib.parse.quote_plus(location_uri, safe='/')
            location_uri_element.appendChild(document.createTextNode(location_uri))
            container.appendChild(location_uri_element)

    def add_file(self, document: Document, container: Element, entry_map: dict[str, EntryInterface], for_nli: bool):
        if for_nli:
            return

        files_list = document.createElement('Files')
        container.appendChild(files_list)

        file_element = document.createElement('File')
        file_element.setAttribute('FileType', 'Text')
        files_list.appendChild(file_element)

        inline_content = document.createElement('InlineContent')
        file_element.appendChild(inline_content)

        inline_content.appendChild(document.createTextNode(str(self.data)))

File: java/src/main/java/com/nuix/edrm/MappingEntry.java
package com.nuix.edrm;

import org.w3c.dom.Document;
import org.w3c.dom.Element;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.OffsetDateTime;
import java.util.HexFormat;
import java.util.LinkedHashMap;
import java.util.Map;

public class MappingEntry extends EntryInterface {
    protected final Map<String, Object> data = new LinkedHashMap<>();
    protected final String parentId;
    protected final String mimeType;

    public MappingEntry(Map<String, Object> mapping, String mimeType) { this(mapping, mimeType, null); }

    public MappingEntry(Map<String, Object> mapping, String mimeType, String parentId) {
        this.parentId = parentId;
        this.mimeType = mimeType;

        Map<String, Object> newData = mapping != null ? mapping : Map.of();
        this.data.putAll(newData);

        fillInitialFields(newData);
        fillGenericFields(mimeType);
    }

    private void fillInitialFields(Map<String, Object> mapping) {
        mapping.forEach((k,v) -> {
            EntryField.Type data_type;
            Object value = v;
            if (v instanceof Boolean) {
                data_type = EntryField.Type.Boolean;
            } else if (v instanceof Short || v instanceof Integer || v instanceof Long) {
                data_type = EntryField.Type.LongInteger;
            } else if (v instanceof Float || v instanceof Double) {
                data_type = EntryField.Type.Decimal;
            } else if (v instanceof OffsetDateTime || v instanceof java.util.Date || v instanceof java.time.LocalDateTime
                    || v instanceof java.time.ZonedDateTime || v instanceof java.time.Instant) {
                data_type = EntryField.Type.DateTime;
            } else {
                data_type = EntryField.Type.Text;
                value = v.toString();
            }

            String key_name = k.trim();
            if (key_name.isEmpty()) {
                int i = 1;
                String placeholder = "key_"+i;
                while (this.fields.containsKey(placeholder)) {
                    placeholder = "key_"+(++i);
                }
                key_name = placeholder;
            }
            putField(FieldFactory.generateField(key_name, data_type, value));
        });
    }

    private void fillGenericFields(String mimeType) {
        putField(FieldFactory.generateField("MIME Type", EntryField.Type.Text, mimeType));
        putField(FieldFactory.generateField("Name", EntryField.Type.Text, getBaseName()));
        
        Map<String, Object> dataCopy = new LinkedHashMap<>(this.data);
        dataCopy.put("name", getBaseName());
        String hashedData = EDRMUtilities.hash_data(dataCopy, "SHA-1");
        putField(FieldFactory.generateField("SHA-1", EntryField.Type.Text, hashedData));
        putField(FieldFactory.generateField("Item Date", EntryField.Type.DateTime, OffsetDateTime.now()));
    }

    public Map<String, Object> getData() {
        return data;
    }

    public String getName() {
        return EDRMUtilities.sanitizeXml(EDRMUtilities.sanitizeFilename( getBaseName()));
    }

    public String getBaseName() {
        String default_rowname_field = EDRMUtilities.EDRM_CONFIG.getOrDefault("default_rowname_field", "Name");
        String name = "";
        if (this.data.containsKey(default_rowname_field)) {
            name = this.data.getOrDefault(default_rowname_field, "").toString();
        } else {
            String nameField = this.data.keySet().stream().filter(k -> k.toLowerCase().contains("name")).findFirst().orElse("");
            if (!nameField.isEmpty()) {
                name = this.data.getOrDefault(nameField, "").toString();
            }
        }

        if (name.isEmpty()) {
            String firstField = this.data.keySet().stream().findFirst().orElse("");
            if (!firstField.isEmpty()) {
                name = this.data.getOrDefault(firstField, "").toString();
            } else {
                name = "MappingEntry";
            }
        }

        return name;
    }

    public String getText() {
        StringBuilder sb = new StringBuilder();
        for (Map.Entry<String, EntryField> entry : fields.entrySet()) {
            if (!sb.isEmpty()) sb.append("\n");
            sb.append(entry.getKey()).append("=").append(entry.getValue().getValue());
        }
        return sb.toString();
    }

    @Override public String getIdentifierField() { return "SHA-1"; }

    @Override public String getTimeField() {
        if (this.data.containsKey(EDRMUtilities.EDRM_CONFIG.getOrDefault("default_itemdate_field", "Item Date"))) {
            return EDRMUtilities.EDRM_CONFIG.getOrDefault("default_itemdate_field", "Item Date");
        }

        String timeField = this.data.keySet().stream().filter(k -> k.toLowerCase().contains("time")).findFirst().orElse("");
        if (!timeField.isEmpty()) {
            return timeField;
        }

        String dateField = this.data.keySet().stream().filter(k -> k.toLowerCase().contains("date")).findFirst().orElse("");
        if (!dateField.isEmpty()) {
            return dateField;
        }

        return null;
    }

    @Override public OffsetDateTime getItemDate() {
        String timeField = this.getTimeField();
        if (null == timeField) return OffsetDateTime.now();

        Object timeValue = this.data.get(timeField);
        if (null == timeValue) return OffsetDateTime.now();
        if (timeValue instanceof String && timeValue.toString().isEmpty()) return OffsetDateTime.now();

        if (timeValue instanceof OffsetDateTime) return (OffsetDateTime)timeValue;
        if (timeValue instanceof java.util.Date) return ((java.util.Date)timeValue).toInstant().atOffset(java.time.ZoneOffset.UTC);
        if (timeValue instanceof java.time.LocalDateTime) return ((java.time.LocalDateTime)timeValue).atOffset(java.time.ZoneOffset.UTC);
        if (timeValue instanceof java.time.ZonedDateTime) return ((java.time.ZonedDateTime)timeValue).toOffsetDateTime();
        if (timeValue instanceof java.time.Instant) return ((java.time.Instant)timeValue).atOffset(java.time.ZoneOffset.UTC);

        if (timeValue instanceof String) {
            try {
                return OffsetDateTime.parse((String)timeValue);
            } catch (Exception e) {
                return OffsetDateTime.now();
            }
        } else {
            return OffsetDateTime.now();
        }
    }

    @Override public String getParentId() { return parentId; }

    @Override public String addAsParentPath(String existingPath) {
        try {
            return URLEncoder.encode(getName(), EDRMUtilities.EDRM_CONFIG
                    .getOrDefault("encoding", StandardCharsets.UTF_8.name())) + "/" + existingPath;
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException(e);
        }
    }

    @Override public void addLocationUri(
            Document document,
            Element container,
            Map<String, EntryInterface> entryMap,
            boolean forNli) {

        if (forNli) {
            Element locationUriElement = document.createElement("LocationURI");
            String locationUri = EDRMUtilities.generateRelativePath(this, entryMap);
            locationUri = URLEncoder.encode(locationUri, StandardCharsets.UTF_8)
                    .replace("+", "%20")
                    .replace("%2F", "/");
            locationUriElement.appendChild(document.createTextNode(locationUri));
            container.appendChild(locationUriElement);
        }
    }

    void serializeFileContent(Document document, Element container, java.util.Map<String, EntryInterface> entryMap) {
        Element externalFile = document.createElement("ExternalFile");
        container.appendChild(externalFile);
        externalFile.setAttribute("FilePath", "natives");
        externalFile.setAttribute("FileName", this.fields.get(this.getIdentifierField()).getValue().toString());

        String md5 = calculateMD5();
        externalFile.setAttribute("Hash", md5);
        externalFile.setAttribute("HashType", "MD5");
    }

    String calculateMD5() {
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            md.update(this.getText().getBytes(StandardCharsets.UTF_8));
            return HexFormat.of().formatHex(md.digest());
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    @Override public void addFile(Document document, Element container, java.util.Map<String, EntryInterface> entryMap, boolean forNli) {
        String text = this.getText();
        if (null == text || text.isEmpty()) return;

        Element filesList = document.createElement("Files");
        container.appendChild(filesList);

        Element file = document.createElement("File");
        file.setAttribute("FileType", forNli ? "Native" : "Text");
        filesList.appendChild(file);

        if(forNli) {
            serializeFileContent(document, file, entryMap);
        } else {
            Element inlineContent = document.createElement("InlineContent");
            inlineContent.appendChild(document.createTextNode(text));
            file.appendChild(inlineContent);
        }
    }
}